### ISSUE TITLE
a better error message when pull occurs on non-existent upstream branch

### ISSUE THREAD

**Issue by tierninho** [desktop/desktop#1325]
I created a repo on dotcom with multiple branches, then cloned it on Desktop. I eventually deleted one of the branches on dotcom, and did a fetch on Desktop. 

As expected, Desktop asked me to publish this branch to dotcom. However, I did a pull on the branch while still in the same status, and I got this error:

`Your configuration specifies to merge with the ref 'refs/heads/tierninho-patch-1'
from the remote, but no such ref was fetched.`

Perhaps this can be rewritten in layman terms for the average end user to comprehend?

<img width="915" alt="screen shot 2017-04-28 at 3 13 22 pm" src="https://cloud.githubusercontent.com/assets/14828183/25549188/533a34c2-2c25-11e7-940c-9b6d3cf4a9e4.png">

v0.0.36, OS X 10.11.5

**Comment by agisilaos** [comment]
👋Would love to start working on it. What's the best approach to start thinking about this issue?
I remember that many error messages aren't defined in the `desktop repo`. 🤔

**Comment by shiftkey** [comment]
> I remember that many error messages aren't defined in the desktop repo. 🤔

@agisilaos they're defined over here - https://github.com/desktop/dugite/blob/master/lib/errors.ts

Adding a new `GitError` enum value, and a `GitErrorRegexes` entry to map the error message to the new `GitError` enum is all that needs to be done there (along with a regression test to validate the change). https://github.com/desktop/dugite/pull/175 is a good example to follow along with.

**Comment by shiftkey** [comment]
Dropping `help wanted` from this issue because I want to call out a workflow that ended up being attempted in #6930 that we didn't expect:

**Along with showing a better error message, the dialog lets the user push the branch again to create it on the remote.**

This felt like a different context to how the user would trigger this problem - when trying to pull in changes from the upstream branch, which now doesn't exist - and a potentially different solution:

**Along with showing a better error message, the dialog lets the user stop tracking the upstream branch.**

Why this approach?

 - Branches can be cleaned up on the GitHub remote as part of normal usage - e.g. after a pull request is merged
 - Branches can also be manually removed through the UI
 - This feels like an opportunity to clean up locally, rather than restore the branch on the remote

In #6930 I suggested that the option here could be to `--unset-upstream` on the current branch, which will put the branch back to a `Publish Branch` state. This feels more explicit and less magical to the end user, if we can get the message right to explain why the remote branch no longer exists.

I also want to point out that pruning merged branches is currently on beta and being tested for GitHub repositories, which will likely reduce the number of users who encounter this flow.
